pyro/distributions/conjugate.py

```diff
@@ -87,6 +93,70 @@ def enumerate_support(self, expand=True):
         return values
 
 
+class DirichletMultinomial(TorchDistribution):
+    r"""
+    Compound distribution comprising of a dirichlet-multinomial pair. The probability of
+    classes (``probs`` for the :class:`~pyro.distributions.Multinomial` distribution)
+    is unknown and randomly drawn from a :class:`~pyro.distributions.Dirichlet`
+    distribution prior to a certain number of Categorical trials given by
+    ``total_count``.
+
+    :param float or torch.Tensor concentration: concentration parameter (alpha) for the
+        Dirichlet distribution.
+    :param int or torch.Tensor total_count: number of Categorical trials.
+    """
+    arg_constraints = {'concentration': constraints.positive, 'total_count': constraints.nonnegative_integer}
+    support = Multinomial.support
+
+    def __init__(self, concentration, total_count=1, validate_args=None):
+        if isinstance(total_count, numbers.Number):
+            total_count = concentration.new_tensor(total_count)
+        total_count_1 = total_count.unsqueeze(-1)
+        concentration, total_count = torch.broadcast_tensors(concentration, total_count_1)
+        total_count = total_count_1.squeeze(-1)
+        self._dirichlet = Dirichlet(concentration)
+        self.total_count = total_count
+        super(DirichletMultinomial, self).__init__(self._dirichlet._batch_shape, validate_args=validate_args)
+
+    @property
+    def concentration(self):
+        return self._dirichlet.concentration
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(DirichletMultinomial, _instance)
+        batch_shape = torch.Size(batch_shape)
+        new._dirichlet = self._dirichlet.expand(batch_shape)
+        new.total_count = self.total_count.expand(batch_shape)
+        super(DirichletMultinomial, new).__init__(batch_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def sample(self, sample_shape=()):
+        probs = self._dirichlet.sample(sample_shape)
+        return Binomial(self.total_count, probs).sample()
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        n = self.total_count
+        alpha = self.concentration
+        alpha_sum = self.concentration.sum(-1)
+        return (_log_factorial(n) + torch.lgamma(alpha_sum) - torch.lgamma(n + alpha_sum) +
+                (torch.lgamma(value + alpha) - _log_factorial(value) - torch.lgamma(alpha)).sum(-1))
```

Comment:
@fritzo just a side note (for us to keep in mind if it helps for performance): when `value` is sparse, we can avoid computing `torch.lgamma(value + alpha) - _log_factorial(value) - torch.lgamma(alpha)` at zero-value indices of `value`.  However, to do so, we have to flatten `value` and `alpha` (to deal with batch_shape), select nonzero indices, compute the result, then transform back to match the batch_shape. I will address this note in the future when facing this distribution in a real case (if it helps).